### ISSUE TITLE
Use HTTPS for all URLs

### ISSUE THREAD

**Issue by aneeshusa** [NixOS/nixpkgs#13809]
I did some searching for previous discussion on this, but it's hard to search for, so let me know if there is some.

We should use HTTPS for all URLs (for URLs that aren't available over HTTPS, we should push the webmasters to adopt it ASAP). Unfortunately, there are a lot of URLs to change (see below).

I'm going to start by working on the generated URLs first, e.g. inside emacs packages, because these take up a huge chunk of the load and can be remedied with little effort. However, there are lot of other URLs that will likely require individual attention, so I thought it would be nice to open a tracking issue to keep tabs on progress.

Is a check for this something we should add to the `CONTRIBUTING.md` or the pull request template?

### Steps to reproduce

```
$ cd workspace/nixpkgs
$ grep -r 'http://' * | wc -l
```

### Expected result

```
0
```

### Actual result

```
24415
```

**Comment by joachifm** [comment]
Some more numbers:
```
% git grep -E 'url = "?http://' pkgs | wc -l
8745
% git grep -E 'homepage = "?http://' pkgs | wc -l
13360
```

**Comment by domenkozar** [comment]
We check anything that comes from the network with a hash. It's way better to fix https://github.com/NixOS/nixpkgs/issues/4491

**Comment by copumpkin** [comment]
Yeah, given that by construction we must check hashes, assuming we have good hashes, the only remaining issue is a potential privacy leak. Not necessarily as big of a deal as integrity, but when people care they generally do care.

**Comment by edolstra** [comment]
I don't see a major reason to require https. Sure, if upstream provides both http and https, then we should prefer https, but I don't think we need to require it.

**Comment by vcunat** [comment]
http -> https transition wouldn't improve privacy too much, as it's still visible who you connect to, when, and how much you download. I believe that's often good enough to determine what files you downloaded exactly.

BTW, https seems much easier to compromise that to find a plausible second pre-image of a given hash. Note that the best *arbitrary*-preimage attack on md5 [has compexity > 2^120](https://en.wiki2.org/wiki/Md5+Milds.5#Preimage_vulnerability). Compare that to the amount of zero-day exploits on common https implementations during the last couple of years...

**Comment by grahamc** [comment]
The sha256 hashes doesn't make HTTPS worthless, and indeed agree it would be valuable to prefer HTTPS. It might be good to add this to a linter, or a PR review checklist (is there such a thing?) and work on it going forward. While I'm not sure updating all existing http URLs is important enough to drop everything and implement, I would bet such a PR would be accepted.

**Comment by aneeshusa** [comment]
Heh, I might have gotten a bit excited after staying up all night because I couldn't sleep.

Agreed that hashes are the main security protection here, and that upgrading them provides a more tangible security boost (I'm waiting eagerly for #12645). Also agreed that the goal is to prefer, not require HTTPS, and that downloading from ipfs and the like is useful. I do think the improved privacy is nice, and we get a small efficiency boost by not having to follow HTTP -> HTTPS redirects on any site with properly-deployed HTTPS.

Also, in my experience the URLs in-tree aren't just used by the Nix tools but by humans, i.e. I'll often click the URLs when looking for package updates or checking out homepages for a package.

I don't know about how to implement linting this, but I'll put in a PR to add this to the PR review checklist.

**Comment by wandernauta** [comment]
It seems that for a lot of packages, HTTP is used because HTTPS just doesn't work (no cert, no valid cert, etc). I've made a [list of URLs in nixpkgs that are currently hardcoded to use HTTP but also work verbatim over HTTPS](https://gist.github.com/wandernauta/f5c87ddda2b702cc56e0), by sending a HEAD request to each and seeing if a 200 OK pops out. It doesn't take redirects, mirrors, etc into account, but it might be a useful place to start if someone wants to put some time into this.

**Comment by zimbatm** [comment]
Changing meta.homepage urls to use https is probably good since it's going to be accessed by humans but please don't spam existing PRs. If you feel strongly about this then it's better submit your own PRs.

For source urls, I believe that link stability is more important than using SSL. We already have a protection mechanism in place that is secure. HTTPS is more computationally expensive, takes longer to establish a connection and is also more likely to fail due to expired certificates and such. I believe solving this is best done by adding another mirror backend like ipfs.

**Comment by spacekitteh** [comment]
Wouldn't it be better to try using https if we are given a http url, and fall back if it doesn't work?

cc @grahamc security, policy discussion

**Comment by vcunat** [comment]
I'm no security expert, but such fallbacks seem to basically kill the security advantages of https. (It's enough to interfere with the https connection attempt and then serve modified http.)

**Comment by spacekitteh** [comment]
true enough.

**Comment by Profpatsch** [comment]
See #34836 and #30906.

**Comment by Infinisil** [comment]
Just found this from @c0bw3b's PR

Want to mention that I recently did a major automated http -> https refactor in https://github.com/NixOS/nixpkgs/pull/42676 for a lot of source urls
